lib/aws-record/record/dirty_tracking.rb

```diff
@@ -0,0 +1,294 @@
+      #  # The change was made in place. Since the String instance representing the vaue of name is the 
+      #  # same as it was originally, the change is not detected.
+      #  model.name_dirty? # => false
+      #  model.name_was    # => 'Alexi'
+      #
+      #  model.name_dirty!
+      #  model.name_dirty? # => true
+      #  model.name_was    # => 'Alexi'
+      #
+      #  model.name << 's'
+      #  model.name        # => 'Alexis'
+      #  model.name_dirty? # => true
+      #  model.name_was    # => 'Alexi'
+      #
+      # @param [String, Symbol] name The name of the attribute to mark as changing.
+      def attribute_dirty!(name)
+        return if attribute_dirty?(name)
+
+        current_value = @data[name]
+
+        @dirty_data[name] = 
+          begin
+            current_value.clone
+          rescue TypeError
+            current_value
+          end
+      end
+
+      # Marks the changes as applied by clearing the current changes and making them accessible through
+      # +previous_changes+.
+      #
+      # # @example
+      #  class Model
+      #    include Aws::Record
+      #    integer_attr :id, hash_key: true
+      #    string_attr  :name
+      #  end
+      #
+      #  model.name   # => 'Alex'
+      #  model.name = 'Nick'
+      #  model.dirty? # => true
+      #
+      #  model.clean!
+      #  model.dirty? # false
+      #
+      def clean!
+        @dirty_data.clear
+      end
+
+      # Returns an array with the name of the attributes with dirty changes.
+      #
+      # @example
+      #  class Model
+      #    include Aws::Record
+      #    integer_attr :id, hash_key: true
+      #    string_attr  :name
+      #  end
+      #
+      #  model.dirty # => []
+      #  model.name  # => 'Alex'
+      #  model.name = 'Nick'
+      #  model.dirty # => ['name']
+      #
+      # @return [Array] The names of attributes with dirty changes.
+      def dirty
+        @dirty_data.keys
+      end
+
+      # Returns +true+ if any attributes have dirty changes, +false+ otherwise.
+      #
+      # @example
+      #  class Model
+      #    include Aws::Record
+      #    integer_attr :id, hash_key: true
+      #    string_attr  :name
+      #  end
+      #
+      #  model.dirty? # => false
+      #  model.name   # => 'Alex'
+      #  model.name = 'Nick'
+      #  model.dirty? # => true
+      #
+      # @return [Boolean] +true+ if any attributes have dirty changes, +false+ otherwise.
+      def dirty?
+        @dirty_data.size > 0
+      end
+
+      # Fetches attributes for this instance of an item from Amazon DynamoDB using its primary key and
+      # the +find(*)+ class method.
+      #
+      # @raise [Aws::Record::Errors::NotFound] if no record exists in the database matching the primary 
+      #  key of the item instance.
+      # 
+      # @return [self] Returns the item instance.
+      def reload!
+        primary_key = self.class.keys.values.inject({}) { |memo, key| memo[key] = send(key); memo }
```

Comment:
Nitpick: Long line, can we break this in to multiple lines of 80 characters or less? Applies to some other lines (esp. comment lines) elsewhere in the file.